src/cppgc.rs

```diff
@@ -0,0 +1,261 @@
+  /// be considered by Heap.
+  pub embedder_id_for_garbage_collected: u16,
+}
+
+impl WrapperDescriptor {
+  pub fn new(
+    wrappable_type_index: InternalFieldIndex,
+    wrappable_instance_index: InternalFieldIndex,
+    embedder_id_for_garbage_collected: u16,
+  ) -> Self {
+    Self {
+      wrappable_type_index,
+      wrappable_instance_index,
+      embedder_id_for_garbage_collected,
+    }
+  }
+}
+
+pub struct HeapCreateParams {
+  wrapper_descriptor: WrapperDescriptor,
+  /// Specifies which kind of marking are supported by the heap.
+  pub marking_support: MarkingType,
+  /// Specifies which kind of sweeping are supported by the heap.
+  pub sweeping_support: SweepingType,
+}
+
+impl HeapCreateParams {
+  pub fn new(wrapper_descriptor: WrapperDescriptor) -> Self {
+    Self {
+      wrapper_descriptor,
+      marking_support: MarkingType::IncrementalAndConcurrent,
+      sweeping_support: SweepingType::IncrementalAndConcurrent,
+    }
+  }
+}
+
+type TraceFn = extern "C" fn(*mut Visitor, *mut ());
+type DestroyFn = extern "C" fn(*mut ());
+
+/// A heap for allocating managed C++ objects.
+///
+/// Similar to v8::Isolate, the heap may only be accessed from one thread at a
+/// time.
+#[repr(C)]
+#[derive(Debug)]
+pub struct Heap(Opaque);
+
+impl Drop for Heap {
+  fn drop(&mut self) {
+    unsafe { cppgc__heap__DELETE(self as *mut Heap) }
+  }
+}
+
+impl Heap {
+  pub fn create(
+    platform: SharedRef<Platform>,
+    params: HeapCreateParams,
+  ) -> UniqueRef<Heap> {
+    let WrapperDescriptor {
+      wrappable_type_index,
+      wrappable_instance_index,
+      embedder_id_for_garbage_collected,
+    } = params.wrapper_descriptor;
+
+    unsafe {
+      UniqueRef::from_raw(cppgc__heap__create(
+        &*platform as *const Platform as *mut _,
+        wrappable_type_index,
+        wrappable_instance_index,
+        embedder_id_for_garbage_collected,
+      ))
+    }
+  }
+
+  pub fn collect_garbage_for_testing(&self, stack_state: EmbedderStackState) {
+    unsafe {
+      cppgc__heap__collect_garbage_for_testing(
+        self as *const Heap as *mut _,
+        stack_state,
+      );
+    }
+  }
+
+  pub fn enable_detached_garbage_collections_for_testing(&self) {
+    unsafe {
+      cppgc__heap__enable_detached_garbage_collections_for_testing(
+        self as *const Heap as *mut _,
+      );
+    }
+  }
+}
+
+/// Base trait for managed objects.
+pub trait GarbageCollected {
+  fn trace(&self, _visitor: &Visitor) {}
+}
+
+/// Members are used to contain strong pointers to other garbage
+/// collected objects. All members fields on garbage collected objects
+/// must be trace in the `trace` method.
+pub struct Member<T: GarbageCollected> {
+  pub handle: *mut (),
+  ptr: *mut T,
+}
+
+impl<T: GarbageCollected> std::ops::Deref for Member<T> {
+  type Target = T;
+
+  fn deref(&self) -> &Self::Target {
+    unsafe { &*self.ptr }
+  }
+}
+
+/// Constructs an instance of T, which is a garbage collected type.
```

Comment:
Let's describe the lifetime of the Box<T> -- ie, when the memory is deallocated, how it may be kept alive, etc.